src/cpp/common/alarm.cc

```diff
@@ -139,18 +147,22 @@ class AlarmImpl : public grpc::internal::CompletionQueueTag {
   EventEngine::TaskHandle callback_timer_handle_ =
       EventEngine::TaskHandle::kInvalid;
   gpr_refcount refs_;
-  grpc_cq_completion completion_;
   // completion queue where events about this alarm will be posted
   grpc_completion_queue* cq_;
   void* tag_;
   std::function<void(bool)> callback_;
+  grpc_core::Notification cancel_notification_;
 };
 }  // namespace internal
 
-Alarm::Alarm() : alarm_(new internal::AlarmImpl()) {}
+Alarm::Alarm() {}
 
 void Alarm::SetInternal(grpc::CompletionQueue* cq, gpr_timespec deadline,
                         void* tag) {
+  if (alarm_ != nullptr) {
+    static_cast<internal::AlarmImpl*>(alarm_)->Destroy();
+  }
+  alarm_ = new internal::AlarmImpl();
```

Comment:
Paying this cost every time we set an alarm is probably going to be expensive, and I don't think we have to. We are concerned about the cancellation path, which can go one of two ways:

1) The EE timer _can_ be cancelled, and the callback will not run. In this case, Alarm state cleanup can be done inline in the `AlarmImpl::Cancel` method, and the callback can be run asynchronously with as little captured state as possible to execute the callback (I think we can get away with just the status). You shouldn't need to wait on any notification, the alarm object can be immediately reused if the asynchronous callback doesn't require alarm state anymore. The same completion-on-the-heap trick you used above should work.

2) The EE timer can't be cancelled, and the callback _will_ be run. In this case, `AlarmImpl::Cancel` can wait on the notification.

On the happy-path with no cancellation, I think we'd pay the cost of one atomic set operation per callback execution.

Reply:
reverted.